### PR TITLE
`tag-selector` iteration

### PR DESCRIPTION
**Pull request by jbwyme**
- Cleaned up a number of UX issues
- Added the ability to select options using the up/down arrow keys
- Factored out list styles to `mp-list-item`
- Added generic function for autocomplete

There are still a few things to do:
- wire loading states back up
- potentially remove backspace behavior
- componentize tags
- fix click outside with nested shadowdom elements

### CODE REVIEW

File: lib/util/string.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/**
```
Comment (jbwyme) [inline lib/util/string.js]: this is copied from IRB so no need to review but we should get IRB updated to use this one

File: test/util/string.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import expect from 'expect.js';
```
Comment (jbwyme) [inline test/util/string.js]: copied from IRB

File: lib/components/list-item/index.jade [hunk 3]
```diff
@@ -0,0 +1,6 @@
+div(class={
+  'mp-list-item-container': true,
+  'mp-list-item-active': $component.isAttributeEnabled('active'),
+  'mp-list-item-static': $component.isAttributeEnabled('static')
```
Comment (rickydoar) [inline lib/components/list-item/index.jade]: Can an item be both active and static? I.e. is there a point to having them be separate attributes instead of just one attribute on vs. off?

File: lib/widgets/tag-selector/index.js [hunk 4]
```diff
@@ -11,41 +12,58 @@ const LOADING_TAG = `loading_tag`;
...
             }
+            case 38: // up arrow
+              this.update({activeTagIndex: Math.max(0, this.state.activeTagIndex - 1)});
+              break;
+            case 40: // down arrow
+              this.update({activeTagIndex: Math.min(this.state.matchingTags.length - 1, this.state.activeTagIndex + 1)});
+              break;
+            default:
+              this._resetActiveTagIndex();
+              break;
```
Comment (rickydoar) [inline lib/widgets/tag-selector/index.js]: so much control 👍

File: lib/widgets/tag-selector/index.js [hunk 5]
```diff
@@ -54,41 +72,86 @@ registerMPElement(`mp-tag-selector`, class extends Component {
...
-    if (attr === `load-state`) {
-      const loadState = this.getAttribute(`load-state`);
-      if (LOAD_STATES.includes(loadState)) {
-        this.update({loadState});
+    if (this.initialized) {
+      switch(attr) {
+        case `load-state`: {
+          const loadState = this.getAttribute(`load-state`);
+          if (LOAD_STATES.includes(loadState)) {
+            //this.update({loadState});
```
Comment (rickydoar) [inline lib/widgets/tag-selector/index.js]: is this supposed to be commented still?

File: lib/components/list-item/index.jade (hunk 3 above)
Comment (jbwyme) [inline lib/components/list-item/index.jade]: Well these are unrelated: something can be both not static and not active. Static just means that hover states don't apply

File: lib/widgets/tag-selector/index.js (hunk 5 above)
Comment (jbwyme) [inline lib/widgets/tag-selector/index.js]: yeah i need to re-think how the loadState stuff is being handled

**Comment by jbwyme** [comment]
Yes, I was under the impression that main-loop queued these for you and batched them. However, there seems to be something related to the multiple update calls and webcomponent.js polyfill. I just tested it using two nested requestAnimationFrame delays and it passes on FF, Safari, and IE whereas it fails with just one.

**Comment by tdumitrescu** [comment]
I think the main difference you're running into is that Chrome fires `attachedCallback` immediately upon inserting the element into the DOM, whereas the others fire it on the next animation frame (probably because of the polyfill, but just possibly because of the different rendering engines). I think there are also now some subtle differences in when `attributeChangedCallback` gets called. In any case, waiting twice on `requestAnimationFrame` for your updates should keep your test deterministic across all browsers, whereas `setTimeout(..., 100)` introduces a race condition and will make the test fail if it can't get through the first frame within 100ms (could happen with e.g. heavy load on hardware running test VM + GC pause).

You're compiling this test case code into ES5 right? In that case you can avoid all the nesting by [`await`ing animation frames](https://github.com/mixpanel/panel/blob/master/test/server/component.js#L31-L41) like we do in panel server-side tests.
